### PR TITLE
OCPBUGS-38292: controller: default to runc when upgrading clusters from 4.17 to 4.18

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Fixes https://issues.redhat.com/browse/OCPBUGS-38292

**- What I did**
This change prevents the automatic switch to crun during the upgrade from version 4.17 to 4.18. Customers can manually opt-in to use crun by deleting the MachineConfig that sets runc as the default runtime.

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: pkg/controller/container-runtime-config/helpers.go
Comment (haircommander) [inline pkg/controller/container-runtime-config/helpers.go:441]: we need to check if crun was already set, and not do this if so. we don't want to clobber

File: pkg/controller/container-runtime-config/container_runtime_config_controller_test.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:106]: hmm. I think this could be more descriptive and actually construct the config map we want to use.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1039]: should be a constant at the top of the file.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1043]: this variable `defaultRuntimeCMIsNotFound` can be deleted and errors.IsNotFound used on line 1044.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1066]: isNotFound variable is probably not needed... just move the IsNotFound inline with the if statement.

File: pkg/controller/container-runtime-config/helpers.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/helpers.go:465]: error isn't handled here... the function should probably return an error.

File: pkg/controller/container-runtime-config/helpers.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/helpers.go:459]: generatedConfigFileList := make([]generatedConfigFile)
and inlining err with a := on line 463 is probably more idiomatic.

**Comment by rphillips** [comment]
Left some comments. We need to run payload upgrade testing on this, and perhaps a clusterbot upgrade to test it out.

File: pkg/controller/container-runtime-config/helpers.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/helpers.go:459]: runc should be the defined constant in the API.

**Comment by rphillips** [comment]
I don't see an edge trigger for the config map. How would a cluster in 4.17 trigger the Add/Update/Delete logic of the controller to create the config map?

**Comment by sohankunkerkar** [comment]
@haircommander @rphillips PTAL

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (haircommander) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1068]: is this case protecting against a ctrcfg already having set default runtime to runc explicitly?

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1068]: yes, I think this case is the `crio-default-container-runtime` already created on the pool in this loop.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1043]:
```suggestion
		return fmt.Errorf("error checking for %s config map: %w", configMapName, err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1065]:
```suggestion
			return fmt.Errorf("error checking for %s machine config: %w", managedKey, err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1075]:
```suggestion
			return fmt.Errorf("could not create crio-default-container-runtime MachineConfig from new Ignition config: %w", err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1079]:
```suggestion
			return fmt.Errorf("error marshalling crio-default-container-runtime config ignition: %w", err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1087]:
```suggestion
			return fmt.Errorf("could not create MachineConfig for crio-default-container-runtime: %w", err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1100]:
```suggestion
		return fmt.Errorf("error creating %s config map: %w", configMapName, err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1112]:
```suggestion
			return nil, fmt.Errorf("could not create MachineConfig from new Ignition config: %w", err)
```

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (rphillips) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:218]: let's create a constant.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:602]: This could be a good candidate to make this a constant.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (kwilczynski) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:600]: Missed one :smile: 

```suggestion
		return fmt.Errorf("failed to create the crio-default-container-runtime MC: %w", err)
```

**Comment by sohankunkerkar** [comment]
The azure-ovn has been a perma-failing test.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (haircommander) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: should this be a ContainerRuntimeConfig instead of a MC?

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: I think this can work fine. managedKey is the name of the mc we are trying to check if it has been created in the current pool. we don't directly create ContainerRuntimeConfig through the API server.

Comment (haircommander) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: @rbaturov may need it in the API server for the NTO to ensure this works, we're talking about it offline now. Also: if we want to avoid an additional reboot update, won't we need to use a ctrcfg to dedup?

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: oh I see, you mean that can avoid the reboot by creating a /etc/crio/crio.conf.d/01-ctrcfg-defaultRuntime for dup configuration.

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: yes, if the configurations in the same file stay the same we don't reboot.

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: If the concern is only aout the extra reboot, instead of creating a ContainerRuntimeConfig, I think in this PR we can write our default configuration to the file `/etc/crio/crio.conf.d/01-ctrcfg-defaultRuntime`, so that a customer-created ContainerRuntimeConfig with the same configuration can avoid draining the node.



> @rbaturov may need it in the API server for the NTO to ensure this works

@haircommander  I am not sure about their usecase, we must ship a containerruntimeconfig CR so they can get it from the API server?

Comment (haircommander) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: yeah the NTO needs to see what the default runtime is to use the same one for the performance profile runtime class

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: I understand their use case is that we just need to ship a default `ContainerRuntimeConfig` CR; we don't have to create the CR or MC in MCO in this PR.

I think it should be similar to the default `ClusterImagePolicy`
we add the manifest of the CR, and MCO can process it by `syncContainerRuntimeConfig` and `syncContainerRuntimeConfigBootstrap`.  Is this the correct approach?

I'm not sure where we should add that manifest.

Comment (haircommander) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: something I'm thinking about is that we can't create the ctrcfg for all branches, we can only do so in 4.17 so we should do it in this PR. otherwise, we'll just update the default runtime in 4.18 when we upgrade, which goes against the aims of this. does that make sense?

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: Yes, I agree that we only need the ContainerRuntimeConfig in 4.17.  I'm just not sure if MCO is the best place to create it, as doing so might cause the sync handler to make a recursive call and process the object on its own. But it would work.

**Comment by rphillips** [comment]
/payload periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-upgrade
/payload periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-upgrade

**Comment by rphillips** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-upgrade
/payload-job periodic-ci-openshift-release-master-ci-4.17-e2e-aws-ovn-upgrade

**Comment by sohankunkerkar** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-upgrade

**Comment by sohankunkerkar** [comment]
@yuqi-zhang could you PTAL?

**Comment by LorbusChris** [comment]
Looks like something is wrong with the referenced Jira ticket. That'll have to be sorted out before this can merge.
Other than that
/lgtm

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:633]: Question: is there a reason why this was removed?

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1062]: I forget if we had this conversation or not, one consideration here is that instead of doing it for all pools, we just do it for master and worker. Any custom pools the user may have will inherit from the worker configuration, so it's just a duplicated entry at the end of the day, and if the user removes the custom pool MC here, the custom pool will undergo no actual change (since the worker config is still taking effect) so unless there's a specific case we're targetting, doing it for just master/worker might be better?

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1042]: Theoretically we try to avoid any direct Get() calls in any of our syncs, as in theory they could trigger and hit API limits.

Practically speaking though, random syncs for containerruntimeconfig objects should not happen very often (~every 10 minutes or so in steady state?) so this is probably fine. It'd be better if we can use a lister, but since this is 4.17 only, I'm fine if we want to keep as is

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1064]: To add to this, creating a containerruntime config object directly is not really an option, since this will conflict with any user-provided ones. CRC objects do not merge like MCs.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:1109]: Similar point to above, maybe we only do this for master/worker pools

File: pkg/controller/container-runtime-config/helpers.go
Comment (yuqi-zhang) [reply pkg/controller/container-runtime-config/helpers.go:462]: I don't think this concern has been addressed?

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:599]: So, this creates `99-$POOL-generated-crio-default-container-runtime` objects, which technically has a potential race condition here. Since:
1. it's not actually generated by the controller (no generated-by-controller-version stamp), and
2. even if it did, the MCO has a blind spot today for new generated MCs by subcontrollers upon upgrades

Such that if the render controller sees that all other MCs had updated, it will generate a new render config before this actually gets to get created, before generating a second render shortly after when this gets created. This isn't too bad since the likely outcome there is an extra reboot for a singular node per pool (not super noticeable but some users have called it out). And really that shouldn't happen unless the first attempt fails due to some blip.

All that's to say we should be aware that this is a rare but possible "unexpected behaviour" but given our options I don't think there's really a much better way to do this.

**Review by yuqi-zhang — COMMENTED** [review]
Left a few comments. I don't think any of them should be considered blocking, but a few nice-to-haves and considerations. If we'd like to merge this ASAP for 4.17 so we can get some soak time I'm not against merging this mostly as is and then fixing anything via a followup (I guess the custom pool behaviour is a bit hard to revert, so curious what your thoughts are on that)

Also @mrunalp this PR also needs an explicit jira/valid-bug (on top of the invalid label you removed) and I don't think I would have that power.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (sohankunkerkar) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:1042]: Discussed this off-line, I'm gonna keep this as-is unless CI quotas are hit.

**Review by yuqi-zhang — APPROVED** [review]
Main concerns are addressed.

**Review by mrunalp — APPROVED** [review]
/lgtm

**Pull request merged**
